### PR TITLE
A45: Exposing OpenCensus Metrics for gRPC retry

### PR DESCRIPTION
**Pull request by dapengzhang0**

### CODE REVIEW

**Comment by dapengzhang0** [comment]
@dfawley @ejona86 @markdroth @yashykt Please review.

**Review by markdroth — CHANGES_REQUESTED** [review]
Thanks for doing this!

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Done.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Fixed.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md:1]: Added a notice in gRFC-A6

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Done.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Done.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Done.

**Review by markdroth — APPROVED** [review]
Looks great!

We'll need to hold off for the full two-week waiting period before merging.

File: A6-client-retries.md
Comment (ejona86) [inline A6-client-retries.md:325]: Use `A45-retry-stats.md` as the link

File: A45-retry-stats.md
Comment (ejona86) [inline A45-retry-stats.md:12]: s/enbaled/enabled/

Comment (ejona86) [inline A45-retry-stats.md:12]: Just use the file name as the link. Ditto elsewhere for gRFC references.

I think I actually need to do the same in my recent gRFC. I only noticed because I saw it recently on another gRFC (Sanjay's?).

File: A45-retry-stats.md
Comment (ejona86) [inline A45-retry-stats.md:16]: The "three additional per-method metrics" link has too many parentheses and doesn't actually render as a link.

File: A45-retry-stats.md
Comment (ejona86) [inline A45-retry-stats.md:38]: This link is also broken because of the parentheses.

File: A45-retry-stats.md
Comment (ejona86) [inline A45-retry-stats.md:63]: How are we going to implement "If retry is not enabled on the client channel"? Are we going to expose whether retries are enabled to tracers?

File: A45-retry-stats.md
Comment (ejona86) [inline A45-retry-stats.md:65]: Is `previous-rpc-attempts` incremented due to a transparent retry? I see it stated later "Transparent retry should also be treated as a distinct RPC with regards to the per-attempt metrics" but not stated here. If so, does that mix well with hedging?

File: A45-retry-stats.md
Comment (ejona86) [inline A45-retry-stats.md:72]: In what metrics are they handled the same way? They aren't included in the retries metric and the transparent_retries metric doesn't include the normal retries.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md:65]: Good question. @markdroth , does the `grpc-previous-rpc-attempts` header in the retry gRFC increment on a transparent retry? IIRC it does not increment on transparent retry.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md:63]: If retry is not enabled, then we keep what we are currently doing. So the users will still see the same view for non-retry case.

Comment (dapengzhang0) [reply A45-retry-stats.md:63]: Yes, we propagate the channel attribute when constructing CensusTracingModule.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md:72]: Not sure if understand your question correctly. They handle the per-attempt metrics in the same way.

File: A6-client-retries.md
Comment (dapengzhang0) [reply A6-client-retries.md]: Done.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Done.

Comment (dapengzhang0) [reply A45-retry-stats.md]: Done.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Dene.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Done

File: A45-retry-stats.md
Comment (markdroth) [inline A45-retry-stats.md:66]: What is the format of the value here?  Is it integer 0 or 1, boolean, or string?

File: A45-retry-stats.md
Comment (ejona86) [reply A45-retry-stats.md:63]: > Yes, we propagate the channel attribute when constructing CensusTracingModule.

That is different than what I said. I want to strongly avoid private information transfer when creating the CensusTracingModule. I was talking about passing the information to the tracer factory when creating a tracer.

File: A45-retry-stats.md
Comment (ejona86) [reply A45-retry-stats.md:72]: It'd probably be more clear if the text was word-wrapped and I didn't have to comment on a full paragraph.

You talk about "treated as distinct RPC" for per-attempt metrics, but I don't actually see anything that is being kept "consistent." If we were consistent, given the description above, we would have a single counter for transparent and non-transparent retries. Instead, we treat transparent retries specially and different. I don't see any metric where non-transparent retries and transparent retries are handled the same way.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md:65]: Java currently is not incrementing this count for transparent retries.

Comment (dapengzhang0) [reply A45-retry-stats.md]: Defined the attribute excluding transparent retry.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Changed to create child span regardless of whether retry is enabled.

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Specified.

**Comment by markdroth** [comment]
This looks great!

File: A45-retry-stats.md
Comment (dapengzhang0) [reply A45-retry-stats.md]: Discussed offline. Clarified a bit and removed misleading commas.

**Pull request merged**
